source/credentials.c

```diff
@@ -199,8 +207,17 @@ struct aws_credentials_provider *aws_credentials_provider_new_chain_default(
 
     environment_provider = aws_credentials_provider_new_environment(allocator, &environment_options);
 
+    bool success = true;
+    struct aws_string *ecs_relative_uri = NULL;
+    aws_get_environment_value(allocator, s_ecs_creds_env_relative_uri, &ecs_relative_uri);
+    struct aws_string *ecs_full_uri = NULL;
+    aws_get_environment_value(allocator, s_ecs_creds_env_full_uri, &ecs_full_uri);
+    struct aws_string *ec2_imds_disable = NULL;
+    aws_get_environment_value(allocator, s_ec2_creds_env_disable, &ec2_imds_disable);
+
     if (environment_provider == NULL) {
-        goto on_error;
+        success = false;
+        goto on_ret;
```

Comment:
move this check up to happen immediately after the `aws_credentials_provider_new_environment()` call. It's weird having the check later after doing a bunch of unrelated stuff.

It's more common to put things at the top of the function that will get set later on (in this case: `success`, and the 3 new ecs strings)

Reply:
Yes, followed into the sense of `define it where you use it`, then due to the `goto` stuff, I have to move them up but was not in the correct place.

Reply:
`define it where you use it` is a good idiom in OTHER languages that automatically destroy objects that go out of scope. The ugly habit of defining everything at the top of a function came from C, where objects don't clean themselves up, so we have to use `goto` and jump to code that cleans everything up, and the things we need to clean up must be defined before our first `goto`.

You should still "define it where you use it" for things like random ints and byte_cursors that don't need clean-up. Only define variables early if you need to clean them up later.

Reply:
Actually, earlier versions of C didn't even let you define variables except at the top of the function, so I guess that's really where the bad habit came from